### PR TITLE
Roll-forward: Beta requirements for JobTrackingWithFinalizers 

### LINKED ISSUE #105200: Tracking with finalizers might count a Pod more than once
This bug is limited to feature gate JobTrackingWithFinalizers.

Investigating #105179, I discovered that sometimes a Pod might be counted more than once.
This could be due to #105199, but there is another failure scenario:

1. Pod A finishes, triggering a sync
2. During sync, Job controller:
    1. Adds UID to `status.uncountedTerminatedPods` status: `{succeeded: 0, uids:[A]}`
    2. Removes finalizer from Pod A
    3. Removes UID from `status.uncountedTerminatedPods` and increases the counter. status: `{succeeded: 1, uids:[]}`
3. We receive the job update event, but the pod informer cache is stale. This triggers a sync.
4. The job controller syncs:
    1. Sees status `{succeeded: 1, uids:[]}` but the pod still has the finalizer.
    2. Adds back UID to `status.uncountedTerminatedPods` status: `{succeeded: 1, uids:[A]}`
    3. Removes finalizer from Pod A (again, but this is noop)
    4. Removes UID from `status.uncountedTerminatedPods` and increases the counter. status: `{succeeded: 2, uids: []}`

/sig apps

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind feature
/sig apps

#### What this PR does / why we need it:

Satisfies beta requirements for JobTrackingWithFinalizers (but it doesn't graduate the feature gate yet):

- Limit the number of pods tracked in a single sync to avoid starvation of small jobs.
- Add metric `job_pod_finished_total`, which increments when the job controller finished tracking a pod (finalizer removed and pod accounted in status counters)

#### Which issue(s) this PR fixes:

Ref kubernetes/enhancements#2307

Fixes #105200

#### Special notes for your reviewer:

This is roll-forward of #104739, reverted in #105181. Commits 2 and 3 are a direct revert of #105181.
The remaining commits include the fixes to the flakiness. The main changes are:
- Doing Pod updates in parallel, reduce number of pods to 50.
- Add an in-memory cache for the pods for which we have removed a finalizer (similar to the creations/deletions expectations).

#### Does this PR introduce a user-facing change?

dropped release note with revert in https://github.com/kubernetes/kubernetes/pull/105181

```release-note
When feature gate JobTrackingWithFinalizers is enabled:
- Limit the number of pods tracked in a single job sync to avoid starvation of small jobs.
- The metric job_pod_finished_total counts the number of finished pods tracked by the job controller
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-tracking-with-finalizers
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/hold
This is still WIP (do not review yet)

**Comment by alculquicondor** [comment]
I will proceed to rerun integration tests a few times to spot flakiness

/test pull-kubernetes-integration

**Comment by alculquicondor** [comment]
3rd run
/test pull-kubernetes-integration

**Comment by alculquicondor** [comment]
4th run 
/test pull-kubernetes-integration

**Comment by alculquicondor** [comment]
5th run
/test pull-kubernetes-integration

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -979,18 +1016,25 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(job *batch.Job, pods []*
...
 // flushed in subsequent calls.
-func (jm *Controller) flushUncountedAndRemoveFinalizers(job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, needsFlush bool) (*batch.Job, bool, error) {
+func (jm *Controller) flushUncountedAndRemoveFinalizers(job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, oldCounters *batch.JobStatus, needsFlush bool) (*batch.Job, bool, error) {
 	var err error
 	if needsFlush {
 		if job, err = jm.updateStatusHandler(job); err != nil {
 			return job, needsFlush, fmt.Errorf("adding uncounted pods to status: %w", err)
 		}
+		recordJobPodFinished(job, *oldCounters)
+		// Shallow copy.
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:1026]: nit: explain why it's sufficient for shallow copy rather than just stating it

**Review by soltysh — APPROVED** [review]
/triage accepted
/priority backlog
/lgtm
/approve

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:1026]: /hold

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go]: Done

**Review by alculquicondor — COMMENTED** [review]
/hold cancel

**Pull request merged**